lib/csv/decoder.ex

```diff
@@ -61,120 +61,102 @@ defmodule CSV.Decoder do
       [%{:x => \"a\", :y => \"b\"}, %{:x => \"c\", :y => \"d\"}]
   """
 
-  def decode(stream, options \\ []) do
-    { headers, stream } = options
-                          |> Keyword.get(:headers, false)
-                          |> get_headers(stream, options)
-
-    row_length = case headers do
-      false -> stream |> get_row_length(options)
-      _ -> headers |> Enum.count
-    end
-
-    num_workers = options
-                  |> Keyword.get(:num_workers, options
-                  |> Keyword.get(:num_pipes, Defaults.num_workers))
-
-    multiline_escape = options
-                        |> Keyword.get(:multiline_escape, true)
+  def decode!(stream, options \\ []) do
+    decode(stream, options)
+    |> handle_errors!
```

Comment:
This is a nice use of composition to make the new API possible - maybe we want to rename `handle_errors!` to `raise_errors!` as well to make the intent more clear?